Service Record information + Description Support + Console Sample

c-michail opened the pull request:
1) Moved service information to Record for serialization purposes 
2) Added Description support
3) Added Console Sample (Like Windows Services Console)

c-michail committed:
Service Record information + Console Sample

1) Moved service information to Record for serialization purposes
2) Added Console Sample

TommiPrami reviewed (commented):
Please rename form and file: Unit23 in 'Unit23.pas' {Form23};

To: SCForm.Console, TSCConsoleForm and SCConsoleForm

TommiPrami reviewed (commented):
Please prefix all local variables with Capital L
var
  Col : Integer;
  ColWidth, CellWidth: Integer;
  Row: Integer;
  
  Like
  
var
  LCol : Integer;
  ...

TommiPrami reviewed (commented):
How about this?

Was thinking that would be more modern to have Helper for the service starte type, migfth be TServiceState, can't remember right now. 

Cells[2, I] := ServiceStateToString(vServiceManager.Services[I].Info.State);

So would leave the ServiceStateToString() but make helper for the type using ServiceStateToString(), so could, if wish, wrrite like this.

Cells[2, I] := vServiceManager.Services[I].Info.State.ToString;

TommiPrami reviewed (commented):
Please don't use variable I in loops.

Use LIndex instead. More descriptive.

TommiPrami reviewed (commented):
I think the delayed start should be added into the: 

  TServiceStartup = (ssAutomatic,
                     ssManual,
                     ssDisabled);
As own ewnum, and then.

ServiceStartupToString()

Would be simpler, because I thin k in Windows it's also own startp type in services manager.

TommiPrami reviewed (commented):
PLease remove ShowMessage, use existing Error Mechanism, or something similar.

  // Update services info
  for LIndex := 0 to GetServiceCount -1 do
  try
    Services[LIndex].RefreshIfNeeded;
  except
    ShowMessage('Erro getting service info: ' + Services[LIndex].FInfo.Name);
  end;
  
  ShowMEssage in Console or Service app, like in Continous Integration in server, is bit problematic. 😉 
  
  Don't ask me how I know 🤣

TommiPrami reviewed (commented):
Please refactor this into it's own helper function. 

  //Get current state of service
  case LServiceInfo.FInfo.Status.dwCurrentState of
    SERVICE_STOPPED: LServiceInfo.FInfo.State := ssStopped;
    SERVICE_START_PENDING: LServiceInfo.FInfo.State := ssStartPending;
    SERVICE_STOP_PENDING: LServiceInfo.FInfo.State := ssStopPending;
    SERVICE_RUNNING: LServiceInfo.FInfo.State := ssRunning;
    SERVICE_CONTINUE_PENDING: LServiceInfo.FInfo.State := ssContinuePending;
    SERVICE_PAUSE_PENDING: LServiceInfo.FInfo.State := ssPausePending;
    SERVICE_PAUSED: LServiceInfo.FInfo.State := ssPaused;
    else
      LServiceInfo.FInfo.State := ssStopped; // Make compiler happy
  end;
  
  Will make code cleaner. And very good candidate of refactgoring. I think.

TommiPrami reviewed (commented):
Plase try to refactor

TServiceInfoRecord

in a way that.

TServiceInfoRecord = record
private
  // Readonly stuff
 public
   property OwnProcess: Boolean read GetOwnProcess;
   property Interactive: Boolean read GetInteractive;
    { Path to the binary that implements the service. }
    property BinaryPathName: string read GetBinaryPathName;
    property Path: string read GetPath;
    property FileName: string read GetFileName;
    property CommandLine: string read GetCommandLine;
    { See what controls the service accepts. }
    property ServiceAccepts: TServiceAccepts read GetServiceAccepts;
    
At lleast thoise, that would not be writeable in Application.

User might think that I can juse modify these. So I would prefer that things that should not be edited, would be read only outside of this unit.

TommiPrami commented:
Sorry that I dumped all this on  you,  I did not realise there was more than I inially thought. And would havbe been way faster me to do the ttrivial changes my self.

But once I started to comment on this pull request didn't realise how stupid it was 🤣

c-michail committed:
Renamed Consoled Sample units

c-michail commented:
> Please prefix all local variables with Capital L var Col : Integer; ColWidth, CellWidth: Integer; Row: Integer;
> 
> Like
> 
> var LCol : Integer; ...

Console Sample is ready with your suggestions (Naming, L variables etc)
I personally user v (aka variable) instead of L.

c-michail commented:
> PLease remove ShowMessage, use existing Error Mechanism, or something similar.
> 
> // Update services info for LIndex := 0 to GetServiceCount -1 do try Services[LIndex].RefreshIfNeeded; except ShowMessage('Erro getting service info: ' + Services[LIndex].FInfo.Name); end;
> 
> ShowMEssage in Console or Service app, like in Continous Integration in server, is bit problematic. 😉
> 
> Don't ask me how I know 🤣

I've used ShowMessage because HandleError brakes the loop when a service doesn't exist. (usually badly uninstall service)
The other way is to silent the exception before looping and restore them back.

```
  // Update services info
  RaiseExceptions := False; // Silent exceptions to avoid breaking the loop on a bad service
  try
    for LIndex := 0 to GetServiceCount -1 do
      Services[LIndex].RefreshIfNeeded;
  finally
    RaiseExceptions := True; // Restore back raising
  end;
```

c-michail commented:
> Sorry that I dumped all this on you, I did not realise there was more than I inially thought. And would havbe been way faster me to do the ttrivial changes my self.
> 
> But once I started to comment on this pull request didn't realise how stupid it was 🤣

No problem mate you need to use the "code" quote above to see the correct formatting.

c-michail committed:
Various improvements

1) Helpers
2) Added TServiceStartup ssAutomaticDelayed
3) Silent exceptions when getting all services information

c-michail commented:
I've done the most you suggested let me know if there something else ...

c-michail commented:
I forgot to add my credits to the units lmao

TommiPrami commented:
Made rest of the changes and some other also.

Both demo apps compile now.

Also removed madExcept stuff from Console app. Will not compile for those that does not have it.

c-michail commented:
Opps i forgot to remove it! Thanks

c-michail commented:
You're renamed TServiceInfo to TService which it might conflict with Vcl.SvcMgr.TService that's why i left it TServiceInfo on purpose.

TommiPrami commented:
Easy to circumvent, by using the unit name as prefix, if needed.

So not really a problem, I think, just a mild annoyance at worst.

-tee-

TommiPrami commented:
> You're renamed TServiceInfo to TService which it might conflict with Vcl.SvcMgr.TService that's why i left it TServiceInfo on purpose.

Prefixed the type names, becaue types like TService and TServiceManaGer, TServiceStatus etrc, are every where.
